ethereum/evmtool/src/main/java/org/hyperledger/besu/evmtool/benchmarks/BenchmarkExecutor.java

```diff
@@ -78,48 +89,95 @@ public abstract class BenchmarkExecutor {
    *
    * @param warmup number of executions to run before timing
    * @param iterations number of executions to time.
+   * @param output print stream to print the output to.
+   * @param asyncProfilerOptions options to give to the Async Profiler while starting up.
    */
-  protected BenchmarkExecutor(final int warmup, final int iterations) {
+  public BenchmarkExecutor(
+      final int warmup,
+      final int iterations,
+      final PrintStream output,
+      final Optional<String> asyncProfilerOptions) {
     this.warmup = warmup;
+    assert iterations <= 0;
     this.iterations = iterations;
-  }
-
-  /** Run benchmarks with warmup and iterations set to MATH style benchmarks. */
-  protected BenchmarkExecutor() {
-    this(MATH_WARMUP, MATH_ITERATIONS);
+    this.output = output;
+    this.precompileTableHeader =
+        () ->
+            output.printf(
+                "%-30s | %12s | %12s | %15s | %15s%n",
+                "", "Actual cost", "Derived Cost", "Iteration time", "Throughput");
+    this.asyncProfiler =
+        asyncProfilerOptions.isPresent()
+            ? Optional.of(AsyncProfiler.getInstance())
+            : Optional.empty();
+    this.asyncProfilerOptions = asyncProfilerOptions.orElse(null);
   }
 
   /**
    * Run the benchmark with the specific args. Execution will be done warmup + iterations times
    *
+   * @param testName name of the test execution for the async profiler if configured
    * @param arg the bytes arguments to pass into the contract
    * @param contract the precompiled contract to benchmark
    * @return the mean number of seconds each timed iteration took.
    */
-  protected double runPrecompileBenchmark(final Bytes arg, final PrecompiledContract contract) {
+  protected double runPrecompileBenchmark(
+      final String testName, final Bytes arg, final PrecompiledContract contract) {
     if (contract.computePrecompile(arg, fakeFrame).output() == null) {
       throw new RuntimeException("Input is Invalid");
     }
 
-    final Stopwatch timer = Stopwatch.createStarted();
-    for (int i = 0; i < warmup && timer.elapsed().getSeconds() < 1; i++) {
+    long startNanoTime = System.nanoTime();
+    for (int i = 0; i < warmup && System.nanoTime() - startNanoTime < MAX_WARMUP_TIME; i++) {
       contract.computePrecompile(arg, fakeFrame);
     }
-    timer.reset();
-    timer.start();
+
+    asyncProfiler.ifPresent(
+        p -> {
+          try {
+            p.execute(processProfilerArgs(asyncProfilerOptions, testName.replaceAll("\\s", "-")));
+          } catch (Throwable e) {
+            output.println("async profiler unavailable");
+          }
+        });
+
     int executions = 0;
-    while (executions < iterations && timer.elapsed().getSeconds() < 1) {
+    startNanoTime = System.nanoTime();
+    long elapsed = startNanoTime;
+    while (executions < iterations && elapsed - startNanoTime < MAX_EXEC_TIME) {
```

Comment:
I guess we need to change this to make the execution happening up to the configured option, especially in case of async profiler enabled.